rclcpp_action/include/rclcpp_action/client.hpp

```diff
@@ -318,31 +325,47 @@ class Client : public ClientBase
     goal_request->goal = goal;
     this->send_goal_request(
       std::static_pointer_cast<void>(goal_request),
-      [this, goal_request, callback, ignore_result, promise](
+      [this,
+      goal_request,
+      goal_response_callback,
+      feedback_callback,
+      result_callback,
+      ignore_result,
+      promise](
         std::shared_ptr<void> response) mutable
       {
         using GoalResponse = typename ActionT::Impl::SendGoalService::Response;
         auto goal_response = std::static_pointer_cast<GoalResponse>(response);
         if (!goal_response->accepted) {
           promise->set_value(nullptr);
+          if (goal_response_callback) {
+            goal_response_callback(nullptr);
+          }
           return;
         }
         GoalInfo goal_info;
         goal_info.goal_id.uuid = goal_request->goal_id.uuid;
         goal_info.stamp = goal_response->stamp;
         // Do not use std::make_shared as friendship cannot be forwarded.
-        std::shared_ptr<GoalHandle> goal_handle(new GoalHandle(goal_info, callback));
+        std::shared_ptr<GoalHandle> goal_handle(
+          new GoalHandle(goal_info, feedback_callback, result_callback));
         if (!ignore_result) {
           try {
             this->make_result_aware(goal_handle);
           } catch (...) {
             promise->set_exception(std::current_exception());
+            if (goal_response_callback) {
+              goal_response_callback(nullptr);
```

Comment:
@jacobperron (meta, can ignore) what the callback gets as its sole argument isn't quite the same as what a wait on the future/promise would get. I wonder if we should be passing the resolved future/promise instead.

Reply:
I have to admit I don't understand what the semantic of passing nullptr here is. What scenario does this occur in?

Reply:
I believe this is the scenario where the "send result" request fails unexpectedly. In retrospect, I'm not sure what the correct behavior is in this case. The goal has already been accepted by the server, but we failed to request the result. Maybe we should attempt to cancel the goal? Maybe we actually return a goal handle, but throw an error if they attempt to get the result?

Reply:
I'm not sure what the right behavior is either, but it should probably be part of the documented behavior of these callbacks.

Reply:
I've updated the signature of the response callback to take a future (20757b4). Now, in this case calling `future::get()` will throw the exception.